### PR TITLE
21506 Legal API - Prioritize businesses eligible for involuntary dissolution by transition filing

### PR DESCRIPTION
**Pull request by AimeeGao**
*Issue #:* /bcgov/entity#21506

*Description of changes:*

- Added sorting logic to prioritize businesses by "transition filing not filed" criterion first, followed by the oldest restoration filing date in ascending order.
- Added and updated unit tests to verify the new sorting logic.

image

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

File: legal-api/tests/unit/services/test_involuntary_dissolution.py [hunk 1]
```diff
@@ -308,3 +308,39 @@ def test_exclude_admin_frozen_businesses(session, test_name, admin_freeze, eligi
 
     check_eligibility, eligibility_details = InvoluntaryDissolutionService.check_business_eligibility(identifier)
     assert check_eligibility == eligible
+
+
+@pytest.mark.parametrize(
```
Comment (eason-pan-bc) [inline legal-api/tests/unit/services/test_involuntary_dissolution.py:313]: Given that there is only one test case, is parameterizing this test necessary? Just a thought!

File: legal-api/src/legal_api/services/involuntary_dissolution.py [hunk 2]
```diff
@@ -141,6 +142,10 @@ def _get_businesses_eligible_query(exclude_in_dissolution=True):
                     _is_limited_restored(),
                     _is_xpro_from_nwpta()
                 )
+            ).\
+            order_by(
+                no_transition_filed_after_restoration.desc(),
+                oldest_restoration_filing_date.asc(),
```
Comment (eason-pan-bc) [inline legal-api/src/legal_api/services/involuntary_dissolution.py:148]: Just wondering, what will happen when there are businesses have the same effective date? then by filing id or random?

File: legal-api/tests/unit/services/test_involuntary_dissolution.py (hunk 1 above)
Comment (AimeeGao) [reply legal-api/tests/unit/services/test_involuntary_dissolution.py:313]: Thanks for the feedback. Mark this PR as a draft to make further updates.

**Comment by AimeeGao** [comment]
Mark this PR as a draft to make further updates. Not ready for review now.

File: legal-api/src/legal_api/services/involuntary_dissolution.py [hunk 3]
```diff
@@ -141,6 +142,12 @@ def _get_businesses_eligible_query(exclude_in_dissolution=True):
                     _is_limited_restored(),
                     _is_xpro_from_nwpta()
                 )
+            ).\
+            order_by(
+                no_transition_filed_after_restoration.desc(),
+                _has_no_transition_filed_after_restoration().asc(),
```
Comment (eason-pan-bc) [inline legal-api/src/legal_api/services/involuntary_dissolution.py:148]: how do we handle the situation when there are multiple businesses have the same effective date? e.g. 
image

Let's say, we are going to select 100 businesses to put into Involuntary Dissolution process, and the 100th - 101st businesses have the same effective date. Then which one should we pick? Or this doesn't actually matter?

Other than these, LGTM

Comment (AimeeGao) [reply legal-api/src/legal_api/services/involuntary_dissolution.py:148]: Thank you for bringing this up, Eason. 

Right now, we don't have a specific requirement for this case, maybe the order between them doesn't matter.

File: legal-api/tests/unit/services/test_involuntary_dissolution.py [hunk 4]
```diff
@@ -308,3 +308,67 @@ def test_exclude_admin_frozen_businesses(session, test_name, admin_freeze, eligi
...
+    # create business that is neither ar_overdue nor transition_overdue
+    business_no_overdue = factory_business(identifier='BC1122334', entity_type=Business.LegalTypes.COMP.value)
+    factory_completed_filing(business_no_overdue, RESTORATION_FILING, filing_type='restoration')
+    factory_completed_filing(business_no_overdue, TRANSITION_FILING_TEMPLATE, filing_type='transition')
+    business_no_overdue.last_ar_date = datetime.utcnow()
+    business_no_overdue.save()
+
+    result = InvoluntaryDissolutionService._get_businesses_eligible_query().all()
+    if exclude:
+        assert not result
```
Comment (chenhongjing) [inline legal-api/tests/unit/services/test_involuntary_dissolution.py:370]: Is `exclude` a necessary parameter in this test?

**Review by chenhongjing — COMMENTED** [review]
Looks good!
Just a comment. 👀

**Review by chenhongjing — APPROVED** [review]
LGTM 😸

**Review by eason-pan-bc — APPROVED** [review]
LGTM 👍

**Comment by argush3** [comment]
can you rebase your latest changes?  

Kevin made some recent updates that affect this code I think.

**Comment by AimeeGao** [comment]
> can you rebase your latest changes?
> 
> Kevin made some recent updates that affect this code I think.

Sure.

File: legal-api/src/legal_api/services/involuntary_dissolution.py [hunk 5]
```diff
@@ -149,6 +150,12 @@ def _get_businesses_eligible_query(eligibility_filters: EligibilityFilters = Eli
...
+                specific_filing_overdue.desc(),
+                _has_specific_filing_overdue().asc()
```
Comment (argush3) [inline legal-api/src/legal_api/services/involuntary_dissolution.py:158]: does this `order_by` code break if any of the clauses returns `None`?

Comment (AimeeGao) [reply legal-api/src/legal_api/services/involuntary_dissolution.py:158]: no_transition_filed_after_restoration and specific_filing_overdue will only return true or false. 

_has_no_transition_filed_after_restoration() might return None when we can't find the expected result. For this case, SQLAlchemy will handle None values by putting them at the end of the result (cause we use ascending order).

The result after ordering would be:
 1. rows sorted based on no_transition_filed_after_restoration. True values appear before False values(In descending order).
 2. The rows with the same no_transition_filed_after_restoration value would be prioritized in ascending order with the Non-None value, and None values will be placed after them.
 3. specific_filing_overdue and _has_specific_filing_overdue have the same ordering logic.
image

Comment (argush3) [reply legal-api/src/legal_api/services/involuntary_dissolution.py:158]: great, thanks for the explanation!

**Review by argush3 — APPROVED** [review]

**Pull request merged**